### ISSUE TITLE
RFC: Plan if custom derives are stabilized

### ISSUE THREAD

**Issue by tomaka** [tomaka/vulkano#256]
If custom derives are stabilized, I plan to make it possible to write something like this:

``` rust
#[macro_use]
extern crate vulkano;
#[macro_use]
extern crate vulkano_macros;

shader!{r#"
#version 450

void main() {
    gl_Position = vec4(1.0, 0.0, 0.0, 1.0);
}

"#}
```

This would work by defining the macro `shader!` in vulkano like this:

``` rust
#[macro_export]
macro_rules! shader {
    ($src:expr) => (
        #[derive(ShaderCompiler)]
        #[src($src)]
        struct Shader;
    )
}
```

And by writing `vulkano_macros` something like this:

``` rust
extern crate rustc_macro;
extern crate vulkano_shaders;

use rustc_macro::TokenStream;

#[rustc_macro_derive(ShaderCompiler)]
pub fn derive_shader_compiler(input: TokenStream) -> TokenStream {
    let input = input.to_string();
    let num_lines = input.lines().count();
    let shader_src = input.lines().skip(1).take(num_lines - 3).collect::<Vec<String>>().concat();

    ... dispatch to vulkano_shaders ...
}
```

Invoking `shader!` would expand to a struct with the shader source code in an attribute, which is then picked up by the handler of `ShaderCompiler` and dispatched to `vulkano_shaders`.

The token stream returned by custom derives actually overwrites the struct, so the user wouldn't get any garbage `struct Shader` in the final code.

**Comment by tomaka** [comment]
Link to tracking issue, for reference: https://github.com/rust-lang/rust/issues/35900

**Comment by tomaka** [comment]
Another idea could be to do this:

```rust
#[derive(Shader)]
#[shader("some_file.glsl")]
struct Shader;
```

Which would build the content of the `Shader` struct and generate some methods.

While for shaders it might be confusing, I think this could be very nice to use for pipeline layout, descriptor sets, and structures defined in GLSL.

For example, in GLSL:

```glsl
struct SomeStruct {
    vec2 a;
    int b;
}
```

And then in Rust:

```rust
#[derive(ExtractFromShader)]
#[shader("shader.glsl")]
struct SomeStruct;
```

Which would be automatically expanded to:

```rust
SomeStruct {
    a: [f32; 2],
    b: i32,
}
```

I like this method a lot, because the user decides exactly which are the types that should be included in scope.

**Comment by tomaka** [comment]
There's a problem however:

```glsl
struct Foo {
    Bar bar[2];
};

struct Bar {
    int a;
};
```

How do you turn `Foo` into Rust code? By assuming that `Bar` is in scope? By automatically generating `Bar` without the user's conscent?

**Comment by nphyx** [comment]
I don't know about the struct extraction issue, but I prefer the latter method for aesthetic and portability reasons: embedding the shader into a sort of heredoc creates complications with editor/IDE integration (IDE has to intelligently detect it to set the correct context for syntax highlighting, autocomplete, etc). 

It also means a lot of shader boilerplate and utility code has to be copypastad rather than vendored, meaning more manual labor keeping it synchronized between projects. I don't know how much of a real-world problem that is (I don't do 3D professionally besides a little WebGL), but it seems like a pain in the ass from my limited perspective.

I can live with hand-writing the Rust side of things until/unless a robust solution is available. Until then it seems like unreliable magic.

**Comment by tomaka** [comment]
Yes, the pros and cons of each solution have to be determined.

Anyway, the latest plan about custom derives in Rust is that they won't be able to mutate the struct. Therefore all the solutions here will need to wait for *real* plugins to be part of Rust.

**Comment by nphyx** [comment]
Is there a way to create the structs using macros rather than mutating them? Something like:
```rust
#[shader("shader.glsl")]
import_shader_struct!("SomeStruct");
```
With shader() creating an intermediary that import_shader_struct can consume? Then if/when Rust gets the necessary features the macro can be deprecated.

**Comment by tomaka** [comment]
> Is there a way to create the structs using macros rather than mutating them?

Yes, with plugins.
We need to wait for them before being able to do anything.

**Comment by tomaka** [comment]
Since macros 1.1 are becoming stable, this could now be done.

**Comment by tomaka** [comment]
The idea is that instead of being a build dependency, vulkano-shaders should be turned into a macros-crate (or whatever they are called).

**Comment by tomaka** [comment]
This can be closed after #359 and #489
